_posts/2023-08-29-grpc-in-scala-with-fs2grpc-and-scalapb.md

```diff
@@ -0,0 +1,966 @@
+        },
+        {
+          name: names[jsonObject["budget"]],
+          quantity: jsonObject["budgetQty"],
+          amount: prices[jsonObject["budget"]],
+        },
+      ],
+    };
+    orderObj.orders.push(jsonFormatted);
+  }
+
+  async function sendForm() {
+    const res = await fetch("http://localhost:8080/submit", {
+      method: "POST",
+      body: JSON.stringify(orderObj),
+    });
+    const resData = await res.json();
+    return resData;
+  }
+  return sendForm()
+    .then((arr) => arr.map((str) => {
+      var p = document.createElement('p')
+      p.textContent = str;
+      p.style.color = "darkorchid";
+      replyDiv.appendChild(p);
+    }))
+    .catch((err) => (replyDiv.innerHTML = err));
+});
+```
+## 7. The main program
+In this section, we bring everything together to run our program. First, we'll need a server to run `restService` that we created previously. Create `Main.scala` in the following path, `src/main/scala/Main.scala`, and add the following code.
+
+```scala
+import cats.effect.*
+import fs2.Stream
+import org.http4s.ember.server.EmberServerBuilder
+import com.comcast.ip4s.*
+import com.xonal.routes.MyRoutes.restService
+
+object Main extends IOApp {
+  def httpServerStream = Stream.eval(
+    EmberServerBuilder
+      .default[IO]
+      .withHost(host"0.0.0.0")
+      .withPort(port"8080")
+      .withHttpApp(restService.orNotFound)
+      .build
+      .use(_ => IO.never)
+  )
+}
+```
+We use `EmberServerBuilder` to create our server with port number 8080 and build it with our `restService`. 
+
+```scala
+import com.xonal.service.Server.grpcServer
+
+object Main extends IOApp {
+  ...
+  val grpcServerStream = Stream.eval(grpcServer)
+
+  def run(args: List[String]): IO[ExitCode] = grpcServerStream
+    .concurrently(httpServerStream)
+    .compile
+    .toList
+    .as(ExitCode.Success)
+}
+```
+We wrap both `httpServerStream` and `grpcServerStream` in an fs2 `Stream` so that we can run them `concurrently` in the last line.
+
+Now Let's test our application.
+When we run our server and navigate to `localhost:8080/index.html`, we should be greeted with the following page.
+
+![index.html](../images/grpc/xonalMobile.png)
+
+Once the form is filled and submitted, we should see a response similar to the following.
+
+![server response](../images/grpc/results.png)
+
+## 8. Managing currency with Squants.
+Throughout the article we've been representing money as `BigDecimal`, we could improve our code by using the `Squants` library to manage currency values, specifically `Squants`' `Money` type. 
+
+Therefore the question arises, if case classes are generated by `Scalapb`, how can we make it so that `Scalapb` generates `Squants`' `Money` type for `amount` and `total`? `Scalapb` provides a mechanism for customising generated types convenient for such scenarios. 
+
+We'll need to add the following two libraries to our `protobuf` project inside `build.sbt`.
+
+```diff
```

Comment:
we already have this set at the beginning of the article, no need for this

Reply:
This wasn't intended to be in the beginning, I was supposed to introduce it here. I removed it from the setting up section